[FLINK-22574] Adaptive Scheduler: Fix cancellation while in Restarting state

rmetzger opened the pull request:
The Canceling state of Adaptive Scheduler was expecting the ExecutionGraph to be in state RUNNING when entering the state.
However, the Restarting state is cancelling the ExecutionGraph already, thus the ExectionGraph can be in state CANCELING or CANCELED when entering the Canceling state.

 Calling the ExecutionGraph.cancel() method in the Canceling state while being in ExecutionGraph.state = CANCELED || CANCELLED is not a problem.

The change is guarded by a new ITCase, as this issue affects the interplay between different AS states.

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java:
```diff
@@ -231,6 +232,43 @@ public void testStopWithSavepointFailOnFirstSavepointSucceedOnSecond() throws Ex
     }
 
+    @Test
+    public void testCancellationOfJobInRestartLoop() throws Exception {
```
Let's give a bit more expressive name. Something like `testJobCancellationWhileRestartingSucceeds` or so.

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java:
```diff
@@ -231,6 +232,43 @@ public void testStopWithSavepointFailOnFirstSavepointSucceedOnSecond() throws Ex
+
+        JobClient client = env.executeAsync();
+
+        NotifyOnRunningAndFailingSource.runningLatch.await();
```
Why do we need this latch here?

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java:
```diff
@@ -231,6 +232,43 @@ public void testStopWithSavepointFailOnFirstSavepointSucceedOnSecond() throws Ex
     }
 
+    @Test
+    public void testCancellationOfJobInRestartLoop() throws Exception {
+        final long timeInRestartingState = 10000L;
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+        // configure a high delay between attempts: We'll stay in RESTARTING for 10 seconds.
+        env.setRestartStrategy(
+                RestartStrategies.fixedDelayRestart(Integer.MAX_VALUE, timeInRestartingState));
+        env.addSource(new NotifyOnRunningAndFailingSource()).addSink(new DiscardingSink<>());
+
+        JobClient client = env.executeAsync();
+
+        NotifyOnRunningAndFailingSource.runningLatch.await();
+
+        // wait until we are in RESTARTING state
+        CommonTestUtils.waitUntilCondition(
+                () -> client.getJobStatus().get() == JobStatus.RESTARTING,
+                Deadline.fromNow(Duration.of(timeInRestartingState, ChronoUnit.MILLIS)),
+                5);
+
+        // now cancel while in RESTARTING state
+        client.cancel().get();
+    }
```
Nit: Does this really have to live in the `flink-tests` module? It feels to me that this should be testable as an integration test in the `flink-runtime` module.

tillrohrmann reviewed (approved):
Thanks for creating this PR @rmetzger. The changes look good to me. I had a couple of smaller comments. +1 for merging after resolving them.

rmetzger commented:
Thanks a lot for the great feedback! I addressed all your comments and will merge the PR once CI is green.
Running the tests in flink-runtime is indeed nicer: and it reduces the runtime from over a second to ~300ms on my machine.

rmetzger commented:
Merging now, there's only one unrelated test failure.
